xdsl/tools/tblgen_to_py.py

```diff
@@ -0,0 +1,642 @@
+                case self._ArgType.OPTIONAL:
+                    fields[name] = f"opt_result_def({constraint})"
+                case self._ArgType.VARIADIC:
+                    fields[name] = f"var_result_def({constraint})"
+                case _:
+                    continue
+
+        for [region, name] in tblgen_op.regions:
+            name = self._resolve_name(name)
+            region = self._get_record(region)
+            variadic = "VariadicRegion" in region.superclasses
+            single_block = (
+                "SizedRegion" in region.superclasses
+                and region.summary == "region with 1 blocks"
+            )
+            match (variadic, single_block):
+                case (False, False):
+                    fields[name] = "region_def()"
+                case (False, True):
+                    fields[name] = 'region_def("single_block")'
+                case (True, False):
+                    fields[name] = "var_region_def()"
+                case (True, True):
+                    fields[name] = 'var_region_def("single_block")'
+                case _:
+                    pass  # Make pyright happy
+
+        for [succ, name] in tblgen_op.successors:
+            name = self._resolve_name(name)
+            succ = self._get_record(succ)
+            if "VariadicRegion" in succ.superclasses:
+                fields[name] = "var_successor_def()"
+            else:
+                fields[name] = "successor_def()"
+
+        field_string = textwrap.indent(
+            "\n\n".join(f"{x} = {d}" for x, d in fields.items()), "    "
+        )
+        string = f'''
+@irdl_op_definition
+class {tblgen_op.name}(IRDLOperation):
+    {"" if tblgen_op.summary == "" else f'"""{tblgen_op.summary}"""'}
+
+{field_string}
+'''
+
+        self.operations[tblgen_op.name] = string
+
+
+def cull_json(output_file: str | None, loader: TblgenLoader):
+    js = loader.js
+    required_fields = {
+        "!name",
+        "!superclasses",
+        "assemblyFormat",
+        "summary",
+        "dialect",
+        "opDialect",
+        "typeName",
+        "attrName",
+        "opName",
+        "arguments",
+        "results",
+        "regions",
+        "successors",
+        "allowedTypes",
+        "bitwidth",
+        "elementType",
+        "valueType",
+        "baseType",
+        "baseAttr",
+        "def",
+        "name",
+    }
+
+    def cull_field(js_in: dict[str, Any]) -> dict[str, Any]:
+        return {key: js_in[key] for key in js_in if key in required_fields}
+
+    culled: dict[str, Any] = {key: cull_field(js[key]) for key in loader.used_records}
+    culled["!instanceof"] = {
+        key: js["!instanceof"][key] for key in ("TypeDef", "AttrDef", "Op", "Dialect")
+    }
+
+    if output_file is not None:
+        with open(output_file, "w") as out_file:
+            print(json.dumps(culled), file=out_file)
+    else:
+        print(json.dumps(culled))
```

Comment:
it feels like the caller should do this dance, and the out parameter here should be `IO[str] | None`